### ISSUE TITLE
Reference.

### ISSUE THREAD

**Issue by Jarod42** [tristanpenman/valijson#5]
I saw that Support for JSON References is in development.
but in the current code I saw that there is:

    boost::optional<typename FetchDocumentFunction<AdapterType>::Type>  fetchDoc

in `populateSchema` and wonder if it can be used and how ?

**Comment by tristanpenman** [comment]
It looks like you have stumbled upon some of the experimental code in Valijson. My intention here is to allow a user to provide a function that will be used to retrieve and load the JSON documents that are referred to in JSON References. The best example of this is in a test file called `test_fetch_document_callback.cpp`.

It is worth noting that this feature is experimental for two reasons. The first is that there are some subtleties around memory management for the adapters returned by the user provided function. I recently changed this to use a boost shared_ptr, but haven't had the time to fully explore the impact of this decision.

The second is that the code to parse the JSON Pointer is quite new, and has not yet been adequately tested.

The take away from this is that the implementation is likely to change, and may even be buggy. I'm working towards a first release of the library, when time allows, but have been a bit busy recently.

**Comment by joris-planorama** [comment]
Thanks for the info.
I'm currently try to validate my schema with the meta-schema which has reference (to itself).
If I understand correctly the code, `fetchDoc` should only be used to retrieve external uri
(so I tried to change the test :

    if (documentUri)

into

    if (documentUri && !documentUri->empty())

so then it seems that this part works.
Unfortunately, the reference of the meta-schema is sort of recursive,
and it seems that your current code doesn't like it (infinite call).
)

**Comment by tristanpenman** [comment]
A quick update on this, for anyone who stumbles across this issue - late last year, I started working on full JSON Reference support on two branches, 'implement-graph-parser' and 'schema-parser-ref-support'. The first branch implements general JSON Reference support through the use of a custom (tiny) JSON parser. I have been calling it a JSON Graph parser because it constructs a graph representation of the underlying JSON documents.

The second branch adds custom support for JSON References to the SchemaParser class. Unfortunately, the JSON Graph parser could not be applied here because the JSON Schema v4 spec includes some subtle changes to the behaviour of JSON References. So this branch revamps the SchemaParser class by directly building in support for JSON References.

Both of these branches were stalled due to some memory management issues that would lead to memory leaks in real world scenarios. But a branch to fix those issues has just been merged into master (in commit ee2d247), so I'll be resuming work on these branches.
